File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    public int solution(int[] numbers, int target) {
        return backTrack(0, 0, numbers, target);
    }
    
    public int backTrack(int depth, int sumValue, int[] numbers, int target) {
        if (depth == numbers.length) {
            if (sumValue == target) {
                return 1;
            }
            return 0;
        }
        
        int addCase = backTrack(depth + 1, sumValue + numbers[depth], numbers, target);
        int minusCase = backTrack(depth + 1, sumValue - numbers[depth], numbers, target);
        return addCase + minusCase;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py


def solution(numbers, target):
    
    def dfs(index, cur_sum):
        
        if index == len(numbers):
            if cur_sum == target:
                return 1
            else:
                return 0
        
        return dfs(index + 1, cur_sum + numbers[index]) + dfs(index + 1, cur_sum - numbers[index])
    
    return dfs(0, 0)
            